diff --git a/tensorflow/lite/kernels/CMakeLists.txt b/tensorflow/lite/kernels/CMakeLists.txt
index 61788660d73..416ea839a0e 100644
--- a/tensorflow/lite/kernels/CMakeLists.txt
+++ b/tensorflow/lite/kernels/CMakeLists.txt
@@ -64,16 +64,16 @@ build_flatbuffers(
 set(DELEGATE_PROVIDERS_SUPP
   ${TFLITE_SOURCE_DIR}/nnapi/sl/SupportLibrary.cc
   ${TFLITE_SOURCE_DIR}/tools/delegates/delegate_provider.cc
-  ${TFLITE_SOURCE_DIR}/tools/evaluation/utils.cc
+  # ${TFLITE_SOURCE_DIR}/tools/evaluation/utils.cc
 )
 
 set(DELEGATE_PROVIDERS
   ${DELEGATE_PROVIDERS_SUPP}
   ${TFLITE_SOURCE_DIR}/tools/delegates/default_execution_provider.cc
   # List of delegates referenced as options in the tensorflow/lite/CMakeLists.txt
-  ${TFLITE_SOURCE_DIR}/tools/delegates/gpu_delegate_provider.cc
+  # ${TFLITE_SOURCE_DIR}/tools/delegates/gpu_delegate_provider.cc
   ${TFLITE_SOURCE_DIR}/tools/delegates/nnapi_delegate_provider.cc
-  ${TFLITE_SOURCE_DIR}/tools/delegates/xnnpack_delegate_provider.cc
+  # ${TFLITE_SOURCE_DIR}/tools/delegates/xnnpack_delegate_provider.cc
 )
 
 if(TFLITE_ENABLE_EXTERNAL_DELEGATE)
@@ -156,7 +156,8 @@ macro(add_kernel_test TEST_SRC TEST_LIB)
     set(DELEGATE_TEST "${TEST_NAME}_delegate")
     add_test(
       NAME ${DELEGATE_TEST}
-      COMMAND cmake -DTEST_EXECUTABLE=$<TARGET_FILE:${TEST_NAME}> -P run-tests.cmake
+      # COMMAND cmake -DTEST_EXECUTABLE=$<TARGET_FILE:${TEST_NAME}> -P run-tests.cmake
+      COMMAND cmake -DTEST_EXECUTABLE=$<TARGET_FILE:${TEST_NAME}> -P ${TFLITE_SOURCE_DIR}/tools/cmake/test_utils/run-tests.cmake
     )
     set_tests_properties(${DELEGATE_TEST} PROPERTIES LABELS "delegate")
   endif()
@@ -164,26 +165,26 @@ endmacro()
 
 # Tests where main() is provided by the file referenced in TEST_FRAMEWORK_MAIN_SRC
 set(TEST_WITH_EXTERNAL_MAIN_LIST
-  internal/averagepool_quantized_test.cc
-  internal/batch_to_space_nd_test.cc
-  internal/conv_per_channel_quantized_16x8_test.cc
-  internal/depthwiseconv_float_test.cc
-  internal/depthwiseconv_per_channel_quantized_16x8_test.cc
-  internal/depthwiseconv_per_channel_quantized_test.cc
-  internal/depthwiseconv_quantized_test.cc
-  internal/log_quantized_test.cc
-  internal/logsoftmax_quantized_test.cc
-  internal/maxpool_quantized_test.cc
-  internal/non_max_suppression_test.cc
-  internal/per_channel_dequantize_test.cc
-  internal/quantization_util_test.cc
-  internal/resize_bilinear_test.cc
-  internal/resize_nearest_neighbor_test.cc
-  internal/softmax_quantized_test.cc
-  internal/strided_slice_logic_test.cc
-  internal/tensor_test.cc
-  internal/tensor_utils_test.cc
-  internal/transpose_utils_test.cc
+  # internal/averagepool_quantized_test.cc
+  # internal/batch_to_space_nd_test.cc
+  # internal/conv_per_channel_quantized_16x8_test.cc
+  # internal/depthwiseconv_float_test.cc
+  # internal/depthwiseconv_per_channel_quantized_16x8_test.cc
+  # internal/depthwiseconv_per_channel_quantized_test.cc
+  # internal/depthwiseconv_quantized_test.cc
+  # internal/log_quantized_test.cc
+  # internal/logsoftmax_quantized_test.cc
+  # internal/maxpool_quantized_test.cc
+  # internal/non_max_suppression_test.cc
+  # internal/per_channel_dequantize_test.cc
+  # internal/quantization_util_test.cc
+  # internal/resize_bilinear_test.cc
+  # internal/resize_nearest_neighbor_test.cc
+  # internal/softmax_quantized_test.cc
+  # internal/strided_slice_logic_test.cc
+  # internal/tensor_test.cc
+  # internal/tensor_utils_test.cc
+  # internal/transpose_utils_test.cc
   acceleration_test_util_internal_test.cc
   activations_test.cc
   add_n_test.cc
diff --git a/tensorflow/lite/tools/cmake/modules/Findgoogletest.cmake b/tensorflow/lite/tools/cmake/modules/Findgoogletest.cmake
index 4fe0b18b040..1f9916da229 100644
--- a/tensorflow/lite/tools/cmake/modules/Findgoogletest.cmake
+++ b/tensorflow/lite/tools/cmake/modules/Findgoogletest.cmake
@@ -22,7 +22,7 @@ include(OverridableFetchContent)
 OverridableFetchContent_Declare(
   googletest
   GIT_REPOSITORY https://github.com/google/googletest.git
-  GIT_TAG release-1.10.0
+  GIT_TAG release-1.12.0
   GIT_SHALLOW TRUE
   GIT_PROGRESS TRUE
   SOURCE_DIR "${CMAKE_BINARY_DIR}/googletest"
diff --git a/tensorflow/lite/tools/cmake/modules/flatbuffers.cmake b/tensorflow/lite/tools/cmake/modules/flatbuffers.cmake
index a9505ed54a6..2a17703c148 100644
--- a/tensorflow/lite/tools/cmake/modules/flatbuffers.cmake
+++ b/tensorflow/lite/tools/cmake/modules/flatbuffers.cmake
@@ -23,7 +23,7 @@ OverridableFetchContent_Declare(
   flatbuffers
   GIT_REPOSITORY https://github.com/google/flatbuffers
   # Sync with tensorflow/third_party/flatbuffers/workspace.bzl
-  GIT_TAG v2.0.6
+  GIT_TAG v2.0.8
   GIT_SHALLOW TRUE
   GIT_PROGRESS TRUE
   SOURCE_DIR "${CMAKE_BINARY_DIR}/flatbuffers"
